### ISSUE TITLE
Add a method to zero-initialize Pool*Arrays

### ISSUE THREAD

**Issue by Demindiro** [godotengine/godot-proposals#2023]
### Describe the project you are working on
I'm working on a project that uses a large `PoolByteArray` for a grid.

### Describe the problem or limitation you are having in your project
I use the elements in the array as an index for another array. However, because `resize()` doesn't zero-initialize the array I get garbage data that causes out of bounds errors.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
One way to fix it would be to have `resize()` zero-initialize the new elements, but this may have a significant performance impact in other code (e.g. code that overwrites the garbage data anyways). Instead I propose that another method is added that can be called if needed.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
I suggest adding `zero_initialize()`*, which simply calls `memset` on the array.

Usage:

```gdscript
var arr := PoolByteArray()
arr.zero_initialize()
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Just don't use it.

### Is there a reason why this should be core and not an add-on in the asset library?
It cannot be done efficiently in GDScript and GDNative is impractical for this purpose.

<sup>\* I don't think this is a good name but I can't think of a better one right now.</sup>

**Comment by LikeLakers2** [comment]
> One way to fix it would be to have resize() zero-initialize the new elements, but this may have a significant performance impact in other code (e.g. code that overwrites the garbage data anyways). Instead I propose that another method is added that can be called if needed.

I haven't looked into it, so please don't take my word for it, but I can't imagine that a zeroing process would have a huge impact on performance unless you're doing it to tens of thousands of elements -- and even then, I'm not sure.

What I would suggest, if we want to make zeroing an option rather than an automatic process, is to instead make it into a new optional parameter, `zero_new_contents` ~~(trademark pending)~~, that defaults to `true`. Then, if someone is getting a noticeable performance impact where it may actually affect gameplay, they can simply add a "false" to the parameters -- the only caveat then is that they'll have to ensure they're not relying on there being zeroes already. This would also work for any weird cases (not judging) where someone might actually be relying on there being random garbage data in the new array slots.

**Comment by Xrayez** [comment]
> I'm working on a project that uses a large `PoolByteArray` for a grid.

I recently implemented `Grid2D` class in Goost, see goostengine/goost#45.

The grid can be resized with `Grid2D.resize(width, height)` method, and it would just resize the internal array basically. It's not a problem for `Variant` because they're "zero-initialized" (`null`) upon resize.

Then there's `Grid2D.fill(value)` which essentially assigns `value` to each element in `Vector<Variant>` (in C++).

So, perhaps it may be worth having the `fill()` method in core for each array type in Godot, because:
1. It's possible to assign any value other than zero.
2. It's much less verbose than `zero_initialize()`.
3. You can fill the array on demand.

**Comment by KoBeWi** [comment]
> If this enhancement will not be used often, can it be worked around with a few lines of script?
Just don't use it.

It can be worked around by just 3 lines:
```
func initialize_array(array):
    for i in array.size():
        array[i] = 0
```
You didn't state why you actually need this method. Unless you create the array multiple times per frame (which is unlikely) the above snippet should be enough in this case.

**Comment by Xrayez** [comment]
> You didn't state why you actually need this method.

It's likely because:

> It cannot be done efficiently in GDScript and GDNative is impractical for this purpose.

I know from experience that massive array manipulations can be quite performance demanding when done via GDScript... But at some point I'd suggest the author to switch to C++ development with Godot if performance becomes a problem more often in a particular project.

Typed arrays may alleviate this problem for `Array` types in Godot 4.0, if I recall correctly.

Also likely because it's a feature which is easy to implement and would be nice to have, because it's general-purpose enough.

**Comment by KoBeWi** [comment]
Well, then it depends on how huge is the huge grid and what is the use.
```
func _ready() -> void:
	var msec = OS.get_ticks_msec()
	var ary = PoolByteArray()
	ary.resize(100000000)
	for i in ary.size():
		ary[i] = 0
	print(OS.get_ticks_msec() - msec)
```
This takes just 6 seconds. If done once (per level), GDScript is efficient enough.

**Comment by dalexeev** [comment]
See godotengine/godot#43033. I prefer the `fill()` method, which allows you to fill an array with an arbitrary value.

**Comment by dalexeev** [comment]
@KoBeWi I strongly support @LikeLakers2's suggestion, because this code

```gdscript
var old_size = arr.size()
arr.resize(new_size)
for i in range(old_size, new_size):
    arr[i] = 0
```

simplifies to


```gdscript
arr.resize(new_size)
```

**Comment by Omicron666** [comment]
i got an optimization for you, so it CAN be done efficiently in gdscript :-P
but i agree, allocating with init value is quite common, and would eventually be be "cleaner"

also testing here efficiency of empty loops which perform quite badly by itself

```
func resize_and_fill(pool, size:int, value=0):
	pool.resize(1)
	pool[0]=value
	while pool.size() << 1 <= size:
		pool.append_array(pool)
	pool.append_array(pool.subarray(0,size-pool.size()-1))
	return pool

const SIZE := 100000000
func _ready() -> void:
	
	var msec 
	
	msec = OS.get_ticks_msec()
	var ary = PoolByteArray()
	ary = resize_and_fill(ary, SIZE, 42) # could be directly ary =  resize_and_fill(PoolByteArray(), SIZE, 42)
	print("ary size: ", ary.size())
	print("timing resize_and_fill: ", OS.get_ticks_msec() - msec)
	#for i in ary.size():
	#	assert(ary[i] == 42)
	
	msec = OS.get_ticks_msec()
	ary = PoolByteArray()
	ary.resize(SIZE)
	for i in ary.size():
		ary[i] = 0
	print("timing init loop: ", OS.get_ticks_msec() - msec)
	
	msec = OS.get_ticks_msec()
	ary = PoolByteArray()
	ary.resize(SIZE)
	for i in range(ary.size()):
		pass
	print("timing empty loop: ", OS.get_ticks_msec() - msec)

	msec = OS.get_ticks_msec()
	ary = PoolByteArray()
	ary.resize(SIZE)
	for i in range(ary.size()):
		pass
	print("timing empty loop with explicite range call: ", OS.get_ticks_msec() - msec)

```
resulting in
```
ary size: 100000000
timing resize_and_fill: 165
timing init loop: 5554
timing empty loop: 2723
timing empty loop with explicite range call: 2788
```

**Comment by Xrayez** [comment]
I think the problem presented in this proposal is already resolved by godotengine/godot#46476 in Godot 4.x, as well as #2181.

godotengine/godot#44827 could directly implement this proposal, but I don't think it's a good solution as it touches core structures which use `CowData` (as noted in https://github.com/godotengine/godot/pull/44827#issuecomment-752932714), and godotengine/godot#46476 handles a lot more use cases in comparison.

**Comment by Calinou** [comment]
This was implemented by https://github.com/godotengine/godot/pull/46476, closing.

(Feel free to open a pull request to backport https://github.com/godotengine/godot/pull/46476 to `3.x` :slightly_smiling_face:)
